### PR TITLE
skip_rule_api

### PR DESCRIPTION
**Pull request by hynguyen2610**
Implemented Skip rule API, waiting for UT

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -24,6 +24,7 @@
 		<java.version>1.8</java.version>
 		<org.mapstruct.version>1.2.0.CR2</org.mapstruct.version>
 		<springfox-swagger.version>2.7.0</springfox-swagger.version>
+
```
Comment (dat-ng) [inline pom.xml]: what is this blank line?

File: src/main/java/com/mgmtp/radio/controller/v1/StationController.java [hunk 2]
```diff
@@ -90,4 +92,11 @@ public StationController(StationService stationService) {
         return stationService.update(id, stationDTO);
     }
 
+    @PutMapping("/update-config/{id}")
```
Comment (dat-ng) [inline src/main/java/com/mgmtp/radio/controller/v1/StationController.java]: PUT method means update, so you should change url

File: src/main/java/com/mgmtp/radio/controller/v1/StationController.java [hunk 3]
```diff
@@ -90,4 +92,11 @@ public StationController(StationService stationService) {
...
+    public Mono<ResponseEntity<StationConfigurationDTO>> updateConfigurationStation(@PathVariable(value = "id") final String id,
```
Comment (dat-ng) [inline src/main/java/com/mgmtp/radio/controller/v1/StationController.java]: Remove ResponseEntity

File: src/main/java/com/mgmtp/radio/controller/v1/StationController.java [hunk 4]
```diff
@@ -90,4 +92,11 @@ public StationController(StationService stationService) {
...
+                                                                                    @Valid @RequestBody final StationConfigurationDTO stationConfigurationDTO) {
+        return stationService.updateConfiguration(id, stationConfigurationDTO)
+            .map(updatedStationConfiguration -> ResponseEntity.ok().body(stationConfigurationDTO))
+            .defaultIfEmpty(new ResponseEntity<>(HttpStatus.NOT_FOUND));
```
Comment (dat-ng) [inline src/main/java/com/mgmtp/radio/controller/v1/StationController.java]: please throw RadioNotFoundException instead return

File: src/main/java/com/mgmtp/radio/domain/station/StationConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.mgmtp.radio.domain.station;
+
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.springframework.data.annotation.Id;
+import org.springframework.data.mongodb.core.mapping.Document;
+
+@NoArgsConstructor
+@Data
+@Document(collection = "stationConfiguration")
```
Comment (dat-ng) [inline src/main/java/com/mgmtp/radio/domain/station/StationConfiguration.java]: name must be ### station_configuration

File: src/main/java/com/mgmtp/radio/dto/skipRule/SkipRuleDTO.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
+package com.mgmtp.radio.dto.skipRule;
+
+import lombok.AccessLevel;
+import lombok.Data;
+import lombok.Setter;
+
+@Data
+public class SkipRuleDTO {
+
+	public static final int BASIC = 0;
```
Comment (dat-ng) [inline src/main/java/com/mgmtp/radio/dto/skipRule/SkipRuleDTO.java]: Should refactor to enum

File: src/main/java/com/mgmtp/radio/dto/skipRule/SkipRuleDTO.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	public static final int ADVANCE = 1;
+
+	int typeId;
+
+    public SkipRuleDTO() {
+    	this.typeId = BASIC;
+    }
+
+	public SkipRuleDTO(final int typeId)  {
```
Comment (dat-ng) [inline src/main/java/com/mgmtp/radio/dto/skipRule/SkipRuleDTO.java]: indentation?

File: src/main/java/com/mgmtp/radio/dto/skipRule/SkipRuleDTO.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+				break;
+			default:
+				description = "Invalid typeId";
+				break;
+		}
+		return description;
+	}
+
+
+//	public class InvalidRuleTypeDtoException extends Exception {
```
Comment (dat-ng) [inline src/main/java/com/mgmtp/radio/dto/skipRule/SkipRuleDTO.java]: Remove comments

File: src/main/java/com/mgmtp/radio/mapper/stationConfiguration/StationConfigurationMapper.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+import com.mgmtp.radio.domain.station.SkipRule;
+import com.mgmtp.radio.domain.station.StationConfiguration;
+import com.mgmtp.radio.dto.skipRule.SkipRuleDTO;
+import com.mgmtp.radio.dto.station.StationConfigurationDTO;
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+@Mapper
+public interface StationConfigurationMapper {
+//	StationConfigurationMapper INSTANCE = Mappers.getMapper(StationConfigurationMapper.class);
```
Comment (dat-ng) [inline src/main/java/com/mgmtp/radio/mapper/stationConfiguration/StationConfigurationMapper.java]: remove comments

File: src/main/java/com/mgmtp/radio/respository/station/StationConfigurationRepository.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package com.mgmtp.radio.respository.station;
+
+import com.mgmtp.radio.domain.station.StationConfiguration;
+import org.springframework.data.mongodb.repository.ReactiveMongoRepository;
+import reactor.core.publisher.Mono;
+
+
+public interface StationConfigurationRepository extends ReactiveMongoRepository<StationConfiguration, String> {
+//	Mono<StationConfiguration> findByIdAndDeletedFalse(String id);
```
Comment (dat-ng) [inline src/main/java/com/mgmtp/radio/respository/station/StationConfigurationRepository.java]: remove comment

File: src/main/java/com/mgmtp/radio/sdo/SubscriptionEvents.java [hunk 11]
```diff
@@ -1,5 +1,6 @@
 package com.mgmtp.radio.sdo;
 
 public enum  SubscriptionEvents {
-    register
+    register,
+    updatePlaylist
```
Comment (dat-ng) [inline src/main/java/com/mgmtp/radio/sdo/SubscriptionEvents.java]: should name update_playlist

File: src/main/java/com/mgmtp/radio/service/station/StationServiceImpl.java [hunk 12]
```diff
@@ -1,23 +1,84 @@
...
-import com.mgmtp.radio.respository.station.StationRepository;
-import org.springframework.stereotype.Service;
-import reactor.core.publisher.Mono;
+import com.mgmtp.radio.domain.station.StationConfiguration;
+import com.mgmtp.radio.dto.station.SongDTO;
+import com.mgmtp.radio.dto.station.StationConfigurationDTO;
 import com.mgmtp.radio.dto.station.StationDTO;
+import com.mgmtp.radio.dto.user.UserDTO;
 import com.mgmtp.radio.mapper.station.StationMapper;
+//import com.mgmtp.radio.mapper.stationConfiguration.StationConfigurationMapper;
```
Comment (dat-ng) [inline src/main/java/com/mgmtp/radio/service/station/StationServiceImpl.java]: remove comment

File: src/main/java/com/mgmtp/radio/service/station/StationServiceImpl.java [hunk 13]
```diff
@@ -1,23 +1,84 @@
...
+			if (station.getOwnerId().equals(user.getId())) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	@AfterReturning(value = "execution(* com.mgmtp.radio.service.station.SongService.downVoteSongInStationPlaylist(..))", returning = "monoSongDTO")
+	public Mono<SongDTO> checkAndSkipSongIfNeeded(Mono<SongDTO> monoSongDTO) {
+		Mono<SongDTO> ms = monoSongDTO.map(songDTO -> {
```
Comment (dat-ng) [inline src/main/java/com/mgmtp/radio/service/station/StationServiceImpl.java]: what is ms mean? Why dont you return here?

File: src/test/java/com/mgmtp/radio/controller/v1/StationControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
+//package com.mgmtp.radio.controller.v1;
+//
```
Comment (dat-ng) [inline src/test/java/com/mgmtp/radio/controller/v1/StationControllerTest.java]: why does this file only contain comment?

File: src/test/java/com/mgmtp/radio/service/station/CreateStationImplTest.java [hunk 15]
```diff
@@ -1,61 +1,67 @@
...
-        StationDTO expected = result.log().block();
-
-//         then
-        assertEquals(stationDTO.getName(), expected.getName());
-        assertEquals(stationDTO.getPrivacy(), expected.getPrivacy());
-        assertEquals(userId, expected.getOwnerId());
-    }
-
-}
+//package com.mgmtp.radio.service.station;
```
Comment (dat-ng) [inline src/test/java/com/mgmtp/radio/service/station/CreateStationImplTest.java]: Comment ?

File: src/main/java/com/mgmtp/radio/controller/v1/StationController.java (hunk 2 above)
Comment (thuytiendang206) [reply src/main/java/com/mgmtp/radio/controller/v1/StationController.java]: We means to update it @dat-ng

File: src/main/java/com/mgmtp/radio/service/station/StationServiceImpl.java (hunk 13 above)
Comment (thuytiendang206) [reply src/main/java/com/mgmtp/radio/service/station/StationServiceImpl.java]: what do you mean, @dat-ng ?  " Why don't you return here?"

**Pull request closed without merging**